helpers/localize.js

```diff
@@ -1,23 +1,24 @@
 import { html } from 'lit';
 import { ifDefined } from 'lit/directives/if-defined.js';
 
-export const acceptedTags = Object.freeze(['d2l-link', 'd2l-tooltip-help', 'p', 'br', 'b', 'strong', 'i', 'em']);
+export const allowedTags = Object.freeze(['d2l-link', 'd2l-tooltip-help', 'p', 'br', 'b', 'strong', 'i', 'em']);
 
-const markupError = `localizeHTML() rich-text replacements must use markup templates with only the following accepted elements: ${acceptedTags}. [link to docs]`;
-const acceptedTagsRegex = new RegExp(`<(?!/?(${acceptedTags.join('|')}))`);
+const markupError = `localizeHTML() rich-text replacements must use markup templates with only the following allowed elements: ${allowedTags}. [link to docs]`;
+const disallowedTagsRegex = new RegExp(`<(?!/?(${allowedTags.join('|')}|>|\s)+).*?>`);
```

Comment:
Okay so parsing out this regex:

- match anything that starts with `<`
- exclude things that optionally have a `/` (closing tag), and have one or more of:
  - the allowed tags OR `>` OR one whitespace
- and then the matching can finish with any number of characters followed by a `>`

I'm wondering: why? Some specific questions:
- why would it be allowed to have more than one of the accepted tags? That means `<pstrongbr>` would be accepted?
- why is `>` in there? That means `<>` and `<p>>` would be accepted?
- why is `\s` in there? That means `<p br> d2l-link>> b>` would be accepted?
- do we still run into excluding things we don't want to? For instance `<d2l-tooltip-help text="< a > are symbols">`? (This will just look at that `< and >` in the text and say it's invalid)

I'm getting more and more reluctant about this approach again 😅. I think it's a pretty good rule to never try to parse XML with regex...

Reply:
Also putting it out there - `<button>` gets past these filters, because `<b.*>` is valid. I think I might understand what you were going for with `>` and `\s`, something like:

```
const disallowedTagsRegex = new RegExp(`<(?!/?((${allowedTags.join('|')})(\\s+|>))).*?>`, 'g');
```

Essentially, saying "any allowed tag, always followed by either `\s+` or `>`"? (Also the double `\\` seems to be required).

Reply:
Ok, made the opening tag optional to exclude things like `<>`, `< >`, and `< s>`:

```javascript
const disallowedTagsRegex = new RegExp(`<(?!/?((${allowedTags.join('|')})?[>\\s])).*?>`);
```

Match anything that:

- Begins with `<`
- Then, is _not_ followed by...
  - Optionally, `/`, then...
  - Optionally, any of the allowed tags
  - Then whitespace or `>`
 - Then anything, any number of times
 - ...until and including `>`
 
<img width="658" alt="Screen Shot 2023-03-14 at 2 13 55 PM" src="https://user-images.githubusercontent.com/1289042/225099684-8cad6140-3343-4226-ad01-0a221b4571ed.png">

And it shouldn't need to be global, since it only needs to match once to be `true`.

Reply:
Because [you can't parse HTML with regex](https://stackoverflow.com/a/1732454), how about a parser-based approach modeled on [this](https://stackoverflow.com/a/47140708)?

```js
const parser = new DOMParser();
const allowedTagsLookup = allowedTags.reduce((acc, v) => { acc[v] = true; return acc; }, {});

function hasDisallowedTags(html) {
	const elements = [ ...parser.parseFromString(html, 'text/html').body.querySelectorAll('*') ];
	return elements.some(v => !allowedTagsLookup[v.tagName.toLowerCase()]);
}
```

I don't like the performance impact of any of this (including the regex approach) given it'll be invoked on each render though. I really wish we could just trust our internal consumers...

Reply:
Scratch that - since this check is being run on the pieces of a template literal, which can be split up like `<a text="${'b'}">` => `['<a text="', '">']`, any actual dom-parser won't work. And neither will a regex that requires a closing `>`.

😭

Reply:
That's fine if we really want to do that, but I would point out that this is not trying to parse HTML with regex. It's simply looking for actually a relatively simple _negative_ pattern. It doesn't care about attributes or child nodes, including text nodes with html entities. It only cares that the tags being used are allowed. Any false negative matches by definition wouldn't create a serious problem (beyond requiring further restriction), and any false positive matches could be worked around by escaping/encoding, which probably should be done anyway.

Reply:
Splitting up html like ```[markup`<a href="`, markup`d2l.com">`]``` doesn't actually render as html because individually they do not parse.

Reply:
Agreed... But ```markup`<a href="${'b'}">link</a>` ```, when parsed by `markup()`, will have an array of strings `['<a href="', '">link</a>']`. And then each of those strings is passed to `validateMarkup()`.

Reply:
Well, in that case the second part contains `</a>` which will fail. I guess someone could do `</${'a'}>` but at that point they're knowingly trying to get around it, which I don't think we want to do anything about.

Reply:
Alright. So we just want to catch any time a consumer uses `<some-tag-name`, unless it's one of our allowed tag names. With the understanding that there will be rare false positives if a consumer does like `text="<3 is a heart"`.

I can bring myself to accept a regex for that. I think it would look like:

```js
const disallowedTagsRegex = new RegExp(`<(?!/?(${allowedTags.join('|')})?([>\\s]|$))`)
```

Reply:
And for `</a>` - okay, but that doesn't apply to a situation like `<input type="${myType}">`, which has no closing tag...

Reply:
Here's what looks like it works best to me:

```javascript
const disallowedTagsRegex = new RegExp(`<(?!/?((${allowedTags.join('|')})?)[\\s>]).+?[>\\s]`)
```

Reply:
Breaking it down:

```
<
(?!
	/?
	(	<- Redundant brackets
		(${allowedTags.join('|')})?
	)
	[\\s>]	<- I think we should include $ to account for string termination
)
.+?[>\\s]	<- I don't think this is necessary, we already account for "<>" and "< " in allowed usage above
```

Reply:
1. Yes, that is redundant
2. `$` doesn't hurt, but isn't necessary since it will just search until the end of the string anyway, and we don't explicitly need it to be the end
3. Yep, this was just helpful for positively matching during testing

Reply:
For 2, this is a false positive without the `$`:

```js
disallowedTagsRegex.test('<d2l-link') // true
disallowedTagsRegex.test('<d2l-link ') // false (expected)
```

Reply:
Okay... maybe `'<d2l-link'` without anything after it won't ever be possible in an `html` template literal.

One other to add to the list of valid terminators though: `/` (e.g. to account for `<br/>`).

Reply:
Ahh, I see what you mean with `$`. It doesn't hurt to include it, so we can.